File: selected_coding_interview/codes/java/lc_876_middle_of_the_linked_list/lc_876_middle_of_the_linked_list.java
/*
* File: lc_876_middle_of_the_linked_list.java
* Created Time: 2025-12-30
* Author: krahets
*/

package lc_876_middle_of_the_linked_list;

import include.*;
import java.util.*;

// ===== Solution Code =====
    class Solution {
        public ListNode middleNode(ListNode head) {
            ListNode fast = head, slow = head;
            while (fast != null && fast.next != null) {
                fast = fast.next.next;
                slow = slow.next;
            }
            return slow;
        }
    }

public class lc_876_middle_of_the_linked_list {
    public static void main(String[] args) {
        // ======= Test Case =======
        // Test case 1: Basic linked list
        ListNode head = ListNode.arrToLinkedList(new int[]{1, 2, 3, 4, 5});

        // ====== Driver Code ======
        Solution slt = new Solution();
        var result = slt.middleNode(head);
        PrintUtil.printLinkedList(result);

    }
}

File: selected_coding_interview/codes/python/lc_876_middle_of_the_linked_list.py
"""
File: lc_876_middle_of_the_linked_list.py
Created Time: 2025-12-30
Author: krahets
"""

from include import *


# ===== Solution Code =====
class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        fast = slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        return slow


# ======= Test Case =======
# Test case 1: Basic linked list
head = list_to_linked_list([1, 2, 3, 4, 5])

# ====== Driver Code ======
slt = Solution()
result = slt.middleNode(head)
print_linked_list(result)
